x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/table/RowInTableRandomizedTests.java

```diff
@@ -0,0 +1,315 @@
+        @Override
+        public void close() {
+            Releasables.close(blocks);
+        }
+    }
+
+    private Table randomTable(BlockFactory blockFactory) {
+        List<List<Object>> keys = new ArrayList<>(maxTableSize);
+        Map<List<Object>, Integer> keyToRow = new HashMap<>(maxTableSize);
+        ElementType[] elementTypes = new ElementType[groups];
+        Block.Builder[] builders = new Block.Builder[groups];
+        try {
+            for (int g = 0; g < groups; g++) {
+                elementTypes[g] = randomFrom(allowedTypes);
+                builders[g] = elementTypes[g].newBlockBuilder(maxTableSize, blockFactory);
+            }
+            for (int k = 0; k < maxTableSize; k++) {
+                List<Object> key = new ArrayList<>(groups);
+                for (int g = 0; g < groups; g++) {
+                    key.add(generators.get(g).gen(elementTypes[g], k));
+                }
+                if (keyToRow.putIfAbsent(key, keys.size()) == null) {
+                    /*
+                     * Duplicate keys aren't allowed in constructors for OffsetInTable
+                     * so just skip them. In most cases we'll have exactly maxTableSize
+                     * entries, but sometimes, say if the generator is `boolean, boolean`
+                     * we'll end up with fewer. That's fine.
+                     */
+                    keys.add(key);
+                    for (int g = 0; g < groups; g++) {
+                        append(builders[g], key.get(g));
+                    }
+                }
+            }
+            return new Table(keys, keyToRow, Block.Builder.buildAll(builders));
+        } finally {
+            Releasables.close(builders);
+        }
+    }
+
+    private record ToLookup(Page rows, List<Set<Integer>> expected) implements Releasable {
+        @Override
+        public void close() {
+            rows.releaseBlocks();
+        }
+    }
+
+    ToLookup toLookup(BlockFactory blockFactory, Table table) {
+        List<Set<Integer>> expected = new ArrayList<>(ROW_COUNT);
+        Block.Builder[] builders = new Block.Builder[groups];
+        try {
+            for (int g = 0; g < groups; g++) {
+                builders[g] = table.blocks[g].elementType().newBlockBuilder(ROW_COUNT, blockFactory);
+            }
+            for (int r = 0; r < ROW_COUNT; r++) {
+                List<List<Object>> generatorKeys = IntStream.range(0, keysPerPosition)
+                    .mapToObj(k -> table.keys.get(between(0, table.keys.size() - 1)))
+                    .toList();
+                for (int g = 0; g < groups; g++) {
+                    List<Object> values = new ArrayList<>(generatorKeys.size());
+                    for (List<Object> key : generatorKeys) {
+                        Object v = key.get(g);
+                        if (v != null) {
+                            values.add(v);
+                        }
```

Comment:
I'm a bit confused why a `null` value for `v` doesn't translate into a `null` added to the builder - won't the builders get misaligned? Could it be that currently nulls don't occur in the keys?

Reply:
Let me go poke the tests some more. `null` is valid key and should get mapped to whatever row has the `null`. And you can look it up. That's how aggs work because that's how postgresql and friends work. Let me double check it.

Reply:
What I've got is actually correct, but it's quite tricky. Tricky in ways ways the certainly deserve a block comment. Adding one.